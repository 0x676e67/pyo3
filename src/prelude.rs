// Copyright (c) 2017-present PyO3 Project and Contributors

//! A collection of items you most likely want to have in scope when working with pyo3
//!
//! The purpose of this module is to alleviate imports of many common pyo3 traits
//! by adding a glob import to the top of pyo3 heavy modules:
//!
//! ```
//! # #![allow(unused_imports)]
//! use pyo3::prelude::*;
//! ```

pub use conversion::{FromPyObject, IntoPyObject, PyTryFrom, PyTryInto, ToPyObject};
pub use err::{PyErr, PyResult};
pub use instance::{AsPyRef, Py, PyToken};
pub use noargs::NoArgs;
pub use object::PyObject;
pub use objectprotocol::ObjectProtocol;
pub use python::Python;
pub use pythonrun::GILGuard;
// This is only part of the prelude because we need it for the pymodinit function
pub use types::PyModule;
// This is required for the constructor
pub use PyRawObject;

pub use pyo3cls::{pyclass, pyfunction, pymethods, pyproto};

#[cfg(Py_3)]
pub use pyo3cls::mod3init as pymodinit;

#[cfg(not(Py_3))]
pub use pyo3cls::mod2init as pymodinit;
