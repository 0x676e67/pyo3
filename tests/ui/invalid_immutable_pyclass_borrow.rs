use pyo3::prelude::*;

#[pyclass(immutable)]
pub struct Foo {
    #[pyo3(get)]
    field: u32,
}

fn borrow_mut_fails(foo: Py<Foo>, py: Python){
    let borrow = foo.as_ref(py).borrow_mut();
}

fn main(){}