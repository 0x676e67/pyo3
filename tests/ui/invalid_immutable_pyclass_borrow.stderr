error[E0277]: the trait bound `Foo: MutablePyClass` is not satisfied
  --> tests/ui/invalid_immutable_pyclass_borrow.rs:10:33
   |
10 |     let borrow = foo.as_ref(py).borrow_mut();
   |                                 ^^^^^^^^^^ the trait `MutablePyClass` is not implemented for `Foo`
